src/app/clusters/water-heater-management-server/water-heater-management-server.cpp

```diff
@@ -0,0 +1,196 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "water-heater-management-server.h"
+
+#include <app/AttributeAccessInterface.h>
+#include <app/AttributeAccessInterfaceRegistry.h>
+#include <app/ConcreteAttributePath.h>
+#include <app/InteractionModelEngine.h>
+#include <app/util/attribute-storage.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::WaterHeaterManagement;
+using namespace chip::app::Clusters::WaterHeaterManagement::Attributes;
+
+using chip::Protocols::InteractionModel::Status;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace WaterHeaterManagement {
+
+CHIP_ERROR Instance::Init()
+{
+    ReturnErrorOnFailure(InteractionModelEngine::GetInstance()->RegisterCommandHandler(this));
+    VerifyOrReturnError(registerAttributeAccessOverride(this), CHIP_ERROR_INCORRECT_STATE);
+
+    return CHIP_NO_ERROR;
+}
+
+void Instance::Shutdown()
+{
+    InteractionModelEngine::GetInstance()->UnregisterCommandHandler(this);
+    unregisterAttributeAccessOverride(this);
+}
+
+bool Instance::HasFeature(Feature aFeature) const
+{
+    return mFeature.Has(aFeature);
+}
+
+// AttributeAccessInterface
+CHIP_ERROR Instance::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    switch (aPath.mAttributeId)
+    {
+    case HeaterTypes::Id:
+        return aEncoder.Encode(mDelegate.GetHeaterTypes());
+    case HeatDemand::Id:
+        return aEncoder.Encode(mDelegate.GetHeatDemand());
+    case TankVolume::Id:
+        if (!HasFeature(Feature::kEnergyManagement))
+        {
+            return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
+        }
+        return aEncoder.Encode(mDelegate.GetTankVolume());
+    case EstimatedHeatRequired::Id:
+        if (!HasFeature(Feature::kEnergyManagement))
+        {
+            return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
+        }
+        return aEncoder.Encode(mDelegate.GetEstimatedHeatRequired());
+    case TankPercentage::Id:
+        if (!HasFeature(Feature::kTankPercent))
+        {
+            return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
+        }
```

Comment:
Overall, it would be best not to have a pull interface unless attributes are FIXED (e.g. TankVolume), but rather a PUSH interface where the delegate calls back-in on change, and the cluster keeps current value. Otherwise you need two sets of methods: one to mark a new value.

Suggest keeping state of mutable attributes (e.g. EatimatedHeatRequired and TankPercentage) state for a given endpoint, and have the delegate push those updates so they can be kept by the Instance and reported properly on *change* to subscriptions.

For the fixed attributes it's fine to have a pull API to get from the delegate.

Reply:
Would it be possible please to point me another cluster where this done?

Reply:
I propose we fix this comment in a follow up PR - but other clusters such as EVSE have done this (arguably wrongly) in the same way as this - i.e. there is a SetXYZAttribute() in the delegate which then calls the MatterReport API, but can see that this proposal may be cleaner to have it in the SDK layer and can be something tested by unit tests.

Reply:
Created https://github.com/project-chip/connectedhomeip/issues/34565 as a follow up.

Will mark this conversation as resolved.